syntax = "proto3";


message Asset {
    message Owner {
        // Public key of the agent who owns the record
        string agent_id = 1;

        // Approximately when the owner was updated, as a Unix UTC timestamp
        uint64 timestamp = 2;
    }

    message Location {
        // Coordinates are expected to be in millionths of a degree
        sint64 latitude = 1;
        sint64 longitude = 2;

        string name = 3;

        // Approximately when the location was updated, as a Unix UTC timestamp
        uint64 timestamp = 4;
    }

    // The user-defined natural key which identifies the object in the
    // real world (for example a serial number)
    string asset_id = 1;

    string asset_type = 2;

    string name = 3;

    // Ordered oldest to newest by timestamp
    repeated Owner owners = 4;
    repeated Location locations = 5;

    string barcode = 6;
    string description = 7;
    
    enum State {
        DAMAGED = 0;
        UNDAMAGED = 1;
        UNKNOWN = 2;
    }

    enum Stage {
        PRODUCTION_STARTED = 0;
        READY_FOR_DISPATCH = 1;
        DISPATCH_FROM_FACTORY = 2;
        RECEIVED_ON_SITE = 3;
        INSTALLED = 4;
    }
    
    State pre_state = 8;
    State post_state = 9;
    Stage stage = 10;
    message PlannedDates {
        uint64 planned_completion_date = 1;
        uint64 planned_delivery_date = 2;
        uint64 planned_installed_date = 3;
    }

    message ActualDates {
        uint64 actual_completion_date = 1;
        uint64 actual_delivery_date = 2;
        uint64 actual_installed_date = 3;
    }
    PlannedDates planned_dates = 11;
    ActualDates actual_dates = 12;
    
    repeated Asset assets = 13;
    message Manufacturer {
        // Public key of the agent who owns the record
        string agent_id = 1;

        // Approximately when the owner was updated, as a Unix UTC timestamp
        uint64 timestamp = 2;
    }

    message Orderer {
        // Public key of the agent who owns the record
        string agent_id = 1;

        // Approximately when the owner was updated, as a Unix UTC timestamp
        uint64 timestamp = 2;
    }

    Manufacturer manufacturer = 14;
    Orderer orderer = 15;


}


message AssetContainer {
    repeated Asset entries = 1;
}
